service/history/workflow/mutable_state_impl.go

```diff
@@ -5840,17 +5840,65 @@ func (ms *MutableStateImpl) UpdateActivity(scheduledEventId int64, updater histo
 	return nil
 }
 
+func (ms *MutableStateImpl) PauseActivityByType(activityType string, identity string, reason string) error {
+	maxPausedActivityTypeCount := ms.config.MutableStateMaxPausedActivityTypeCount(ms.namespaceEntry.Name().String())
+	maxPausedActivityTypeLength := ms.config.MutableStateMaxPausedActivityTypeLength()
+
+	if len(activityType) == 0 {
+		return serviceerror.NewInvalidArgument("activity type name is empty")
+	}
+	if len(activityType) > maxPausedActivityTypeLength {
+		return serviceerror.NewResourceExhausted(enumspb.RESOURCE_EXHAUSTED_CAUSE_PERSISTENCE_STORAGE_LIMIT, fmt.Sprintf("activity type name is too long: %s", activityType))
+	}
+
+	if ms.executionInfo.PauseInfo == nil {
+		ms.executionInfo.PauseInfo = &persistencespb.WorkflowPauseInfo{
+			ActivityPauseInfos: []*persistencespb.ActivityPauseInfo{},
+		}
+	}
+
+	pausedActivities := ms.executionInfo.PauseInfo.ActivityPauseInfos
+
+	// check if the activity is already paused
+	if slices.ContainsFunc(pausedActivities, func(activity *persistencespb.ActivityPauseInfo) bool {
+		return activity.ActivityType == activityType
+	}) {
+		return nil
```

Comment:
Can we use the request ID as we discussed in the design and only ignore duplicate requests for the same ID (e.g. a retry)?

Reply:
Yes. We need a proto change for that. Will do it after that. I noted the same in the PR description.